pkg/account/account.go

```diff
@@ -0,0 +1,237 @@
+		return err
+	}
+	return nil
+}
+
+// Validate the account data
+func (a *Account) Validate() error {
+	err := validation.ValidateStruct(&a.data,
+		validation.Field(&a.data.AdminRoleArn, validation.NotNil),
+		validation.Field(&a.data.ID, accountIDRules...),
+		validation.Field(&a.data.LastModifiedOn, validation.NotNil),
+		validation.Field(&a.data.Status, validation.NotNil),
+		validation.Field(&a.data.CreatedOn, validation.NotNil),
+		validation.Field(&a.data.PrincipalRoleArn, validation.NilOrNotEmpty),
+		validation.Field(&a.data.PrincipalPolicyHash, validation.NilOrNotEmpty),
+	)
+	if err != nil {
+		return errors.NewValidation("account", err)
+	}
+	return nil
+}
+
+// Update the Account record in DynamoDB
+func (a *Account) Update(d model.Account, am Manager) error {
+	err := validation.ValidateStruct(&d,
+		// ID has to be empty
+		validation.Field(&d.ID, validation.NilOrNotEmpty, validation.In(*a.data.ID)),
+		validation.Field(&d.AdminRoleArn, validation.By(isNilOrUsableAdminRole(am))),
+		validation.Field(&d.ID, validation.By(isNil)),
+		validation.Field(&d.LastModifiedOn, validation.By(isNil)),
+		validation.Field(&d.Status, validation.By(isNil)),
+		validation.Field(&d.CreatedOn, validation.By(isNil)),
+		validation.Field(&d.PrincipalRoleArn, validation.By(isNil)),
+		validation.Field(&d.PrincipalPolicyHash, validation.By(isNil)),
+	)
+	if err != nil {
+		return errors.NewValidation("account", err)
+	}
+
+	if d.AdminRoleArn != nil {
+		a.data.AdminRoleArn = d.AdminRoleArn
+	}
+	if d.Metadata != nil {
+		a.data.Metadata = d.Metadata
+	}
+
+	err = a.save()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Delete finds a given account and deletes it if it is not of status `Leased`. Returns the account.
+func (a *Account) Delete() error {
+
+	err := validation.ValidateStruct(&a.data,
+		validation.Field(&a.data.Status, validation.NotNil, validation.By(isAccountNotLeased)),
+	)
+	if err != nil {
+		return errors.NewConflict("account", *a.data.ID, err)
+	}
+
+	err = a.writer.DeleteAccount(&a.data)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetAccountByID returns an account from ID
+func GetAccountByID(ID string, d SingleReader, wd WriterDeleter) (*Account, error) {
+
+	newAccount := Account{
+		writer: wd,
+	}
+	data, err := d.GetAccountByID(ID)
+	if err != nil {
+		return nil, err
+	}
+	newAccount.data = *data
+
+	return &newAccount, err
+}
+
+// New returns an account from ID
+func New(wd WriterDeleter, data model.Account) *Account {
+	now := time.Now().Unix()
+	account := &Account{
+		writer: wd,
+		data:   data,
+	}
+	account.data.CreatedOn = &now
+	account.data.LastModifiedOn = &now
+	return account
+}
+
+// MarshalJSON Marshals the data inside the account
+func (a *Account) MarshalJSON() ([]byte, error) {
+	return json.Marshal(a.data)
+}
+
+// GetReadyAccount returns an available account record with a
+// corresponding status of 'Ready'
+func GetReadyAccount(d Reader, wd WriterDeleter) (*Account, error) {
```

Comment:
Could we pass in a single `ReaderWriterDeleter` object here?

Reply:
I was thinking about that but we only keep the wd and it isn't easy to split apart.  The `wd` was a fallout on removing it from being passed into so many functions.

The other option would be to have a Setter type function that could be used to set it when needed.

After discussion I'm going to remove this function.